packages/celotool/src/e2e-tests/validator_handshake_tests.ts

```diff
@@ -0,0 +1,263 @@
+import { eqAddress } from '@celo/utils/lib/address'
+import { assert } from 'chai'
+import _ from 'lodash'
+import sleep from 'sleep-promise'
+import Web3 from 'web3'
+import { Admin } from 'web3-eth-admin'
+import { GethRunConfig } from '../lib/interfaces/geth-run-config'
+import { GethInstanceConfig } from '../lib/interfaces/geth-instance-config'
+import { configureInstances, getContext, jsonRpc, killInstance, restartInstance } from './utils'
+
+const VALIDATORS = 3
+const PROXIED_VALIDATORS = 1
+const EPOCH = 20
+
+const TMP_PATH = '/tmp/e2e'
+
+describe('Validator handshake tests', () => {
+  const gethConfig: GethRunConfig = {
+    networkId: 1101,
+    network: 'local',
+    runPath: TMP_PATH,
+    migrate: false,
+    instances: getInstances(),
+    genesisConfig: {
+      epoch: EPOCH,
+    },
+  }
+  const context: any = getContext(gethConfig)
+
+  before(async function(this: any) {
+    this.timeout(0)
+    await context.hooks.before()
+  })
+
+  after(context.hooks.after)
+
+  describe('Validator handshake', () => {
+    before(async function() {
+      // reset the instances for each test
+      gethConfig.instances = getInstances()
+      configureInstances(gethConfig)
+      this.timeout(0)
+      await context.hooks.restart()
+    })
+
+    it('allows a validator with a new enode to identify itself to a newly peered validator whose maxpeers is 0', async function(this: any) {
+      this.timeout(20000) // 20 seconds
+      // If a consensus round fails during this test, the results are inconclusive.
+      // Retry up to two times to mitigate this issue. Restarting the nodes is not needed.
+      this.retries(2)
+
+      const val0Provider = `http://localhost:${gethConfig.instances[0].rpcport}`
+      const val0Web3 = new Web3(val0Provider)
+      const val0Admin = new Admin(val0Provider)
+      const val0InfoBefore = await val0Admin.getNodeInfo()
+      const val0EnodeBefore = val0InfoBefore.enode
+
+      const val1Provider = `http://localhost:${gethConfig.instances[1].rpcport}`
+      const val1Web3 = new Web3(val1Provider)
+      const val1Admin = new Admin(val1Provider)
+
+      // restart instance 1 with maxPeers as 0
+      gethConfig.instances[1].maxPeers = 0
+      await restartInstance(gethConfig, gethConfig.instances[1])
+
+      // restart validator 0, and give it a new random nodekey
+      await restartInstance(gethConfig, gethConfig.instances[0], true)
+
+      const val0InfoAfter = await val0Admin.getNodeInfo()
+      const val0EnodeAfter = val0InfoAfter.enode
+
+      assert(
+        !_.isEqual(val0EnodeBefore, val0EnodeAfter),
+        'Restarting instance 0 with a new nodekey did not result in a new enode'
+      )
+
+      // validator 0 still has its valEnodeTable, so it will try to connect to
+      // validator 1 immediately. It should identify itself as a validator and
+      // immediately be accepted as a ValidatorPurpose peer by validator 1
+      const val1Peers = await val1Admin.getPeers()
```

Comment:
Could this be the result of val1 connecting to the bootnode, gossiping the announce messages, and having the connection be established after the val0 node received the latest announce?

Reply:
This test doesn't have `useBootnode: true` in the `GethRunConfig`, and because validator 0 will have a new enode, none of the other nodes will be able to initiate a connection to validator 0.

As an extra check to make sure we can rely on the val enode table, I added some code to ensure that the val enode table of validator 0 contains the correct enode of validator 1 after validator 0 has been restarted. There's also a check at the end of the test to make sure that the connection between validator 0 and validator 1 was initiated by validator 0, which confirms that it was able to get around validator 1's max peer check.